pkg/types/namespacedname.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+// NamespacedName comprises a resource name, with a mandatory namespace,
+// rendered as "<namespace>/<name>".  Being a type captures intent and
+// helps make sure that UIDs, namespaced names and non-namespaced names
+// do not get conflated in code.  It also provides a convenient place for
+// utility functions to do things like splitting, joining and rendering.
+type NamespacedName struct {
+	namespace string
+	name      string
+}
+
+// NewNamespacedName is the constructor
+func NewNamespacedName(namespace, name string) *NamespacedName {
```

Comment:
What is an "invalid string"?  While it is somewhat EASIER to say NamespacedName{Name: "foo"} than NewNamespacedName("foo", ""), it's still possible in both modes.

There's something about Go that took me a while to adapt to - there's this sense that encapsulation is not particularly idiomatic.  Hiding fields and wrapping them in accessors is just not the done thing in Go.  We can only protect people from themselves so far - omitting a field in a 2-field tuple struct is a programmer error.  I don't think we want to be doing data validation every time we construct one of these, so maybe we should just stop pretending that we're protecting anyone?

If you want to stick with this model, I'll approve, but on reading it, I would do it the simpler way, myself.